### PR TITLE
feat: solving magic sequence problem with backtracking

### PR DESCRIPTION
**Pull request by Jxtopher**
#### Description of Change

Solving magic sequence problem with a backtracking.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1533"></a>

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -2,5 +2,63 @@
      "C_Cpp.clang_format_style": "{ BasedOnStyle: Google, UseTab: Never, IndentWidth: 4, TabWidth: 4, AllowShortIfStatementsOnASingleLine: false, IndentCaseLabels: true, ColumnLimit: 80, AccessModifierOffset: -3, AlignConsecutiveMacros: true }",
      "editor.formatOnSave": true,
      "editor.formatOnType": true,
-     "editor.formatOnPaste": true
+     "editor.formatOnPaste": true,
+     "files.associations": {
+          "cctype": "cpp",
```
Comment (Panquesito7) [inline .vscode/settings.json:7]: Hmm, I'm not sure if this file should be modified. @ayaankhan98, @mishraabhinn, thoughts?

File: backtracking/magic_sequence.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/*
+ * @brief [Magic sequence](https://www.csplib.org/Problems/prob019/)
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:2]:
```suggestion
 * @brief [Magic sequence](https://www.csplib.org/Problems/prob019/) implementation
```

File: backtracking/magic_sequence.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * @details
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:4]: Please add a detailed description of the algorithm.

File: backtracking/magic_sequence.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * @author [Jxtopher](https://github.com/jxtopher)
+ */
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <list>
+#include <numeric>
+#include <vector>
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:14]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Your functions and namespaces are missing documentation. Please check the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program for how to make your code properly.

File: .vscode/settings.json (hunk 1 above)
Comment (mishraabhinn) [reply .vscode/settings.json:7]: I don't have an idea about this @Panquesito7

Comment (Jxtopher) [reply .vscode/settings.json]: Indeed the ".vscode / settings.json" file must not be modified. The CI script returned the file to its original state.

Comment (ayaankhan98) [reply .vscode/settings.json]: @Jxtopher you deleted the file, please revert it back with no changes as it was originally.

File: backtracking/magic_sequence.cpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                                                        0, 1, 0, 0, 0};
+    assert(backtracking::magic_sequence::is_magic(s_magic));
+
+    backtracking::magic_sequence::sequence_t s_not_magic = {5, 2, 1, 0, 0,
+                                                            0, 1, 0, 0, 0};
+    assert(!backtracking::magic_sequence::is_magic(s_not_magic));
+}
+
+int main() {
+    test();
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:121]:
```suggestion
    test();  // run self-test implementations
```

File: backtracking/magic_sequence.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+ * 
+ */
+static void test() {
+    backtracking::magic_sequence::sequence_t s_magic = {6, 2, 1, 0, 0,
+                                                        0, 1, 0, 0, 0};
+    assert(backtracking::magic_sequence::is_magic(s_magic));
+
+    backtracking::magic_sequence::sequence_t s_not_magic = {5, 2, 1, 0, 0,
+                                                            0, 1, 0, 0, 0};
+    assert(!backtracking::magic_sequence::is_magic(s_not_magic));
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:117]:
```suggestion
    backtracking::magic_sequence::sequence_t s_magic = {6, 2, 1, 0, 0,
                                                        0, 1, 0, 0, 0};
    assert(backtracking::magic_sequence::is_magic(s_magic));
    backtracking::magic_sequence::sequence_t s_not_magic = {5, 2, 1, 0, 0,
                                                            0, 1, 0, 0, 0};
    assert(!backtracking::magic_sequence::is_magic(s_not_magic));
```

File: backtracking/magic_sequence.cpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+int main() {
+    test();
+
+    for (unsigned int i = 2; i < 12; i++) {
+        std::cout << "Solution for n = " << i << std::endl;
+        std::list<backtracking::magic_sequence::sequence_t> r1;
+        backtracking::magic_sequence::sequence_t s1(i, i);
+        backtracking::magic_sequence::solve(&s1, &r1);
+        for (const auto& item : r1) backtracking::magic_sequence::print(item);
+    }
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:129]: Could you please add a bit of documentation of what you're doing here?

File: backtracking/magic_sequence.cpp (hunk 6 above)
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:117]: Could you please add a bit of documentation of what you're doing here?

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work!

File: backtracking/magic_sequence.cpp (hunk 7 above)
Comment (Jxtopher) [reply backtracking/magic_sequence.cpp:127]: Yes. I propose : solve magic sequences of size 2 to 11 and print the solutions.

File: backtracking/magic_sequence.cpp (hunk 6 above)
Comment (Jxtopher) [reply backtracking/magic_sequence.cpp]: Yes. thanks for your question.

File: backtracking/magic_sequence.cpp (hunk 7 above)
Comment (Jxtopher) [reply backtracking/magic_sequence.cpp:127]: Yes. Thanks for your question.

File: backtracking/magic_sequence.cpp [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+ * @brief [Magic sequence](https://www.csplib.org/Problems/prob019/) implementation
+ *
+ * @details Solve the magic sequence problem with a backtraking
+ * 
+ * "A magic sequence of length $n$ is a sequence of integers $x_0 
+ *  \ldots x_{n-1}$ between $0$ and $n-1$, such that for all $i$ 
+ *  in $0$ to $n-1$, the number $i$ occurs exactly $x_i$ times in 
+ * the sequence. For instance, $6,2,1,0,0,0,1,0,0,0$ is a magic 
+ * sequence since $0$ occurs $6$ times in it, $1$ occurs twice, etc." 
+ * Quote of https://www.csplib.org/Problems/prob019/
```
[the remark this answers is not in the document]
Comment (Jxtopher) [reply backtracking/magic_sequence.cpp]: I think it should be indicated that this is a quote

**Comment by Jxtopher** [comment]
Thank you for these improvements :)

File: backtracking/magic_sequence.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Panquesito7) [reply backtracking/magic_sequence.cpp]: > I think it should be indicated that this is a quote

In other files, there's no link provided to the quote.

[the remark this answers is not in the document]
Comment (Jxtopher) [reply backtracking/magic_sequence.cpp]: I understand, I think the sources should always be cited, out of respect for the author. Here, it's copy-paste and not a rewrite of the definition.

File: backtracking/magic_sequence.cpp [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+ * implementation
+ *
+ * @details Solve the magic sequence problem with a backtraking
+ *
+ * "A magic sequence of length $n$ is a sequence of integers $x_0
+ *  \ldots x_{n-1}$ between $0$ and $n-1$, such that for all $i$
+ *  in $0$ to $n-1$, the number $i$ occurs exactly $x_i$ times in
+ * the sequence. For instance, $6,2,1,0,0,0,1,0,0,0$ is a magic
+ * sequence since $0$ occurs $6$ times in it, $1$ occurs twice, etc."
+ * Quote of https://www.csplib.org/Problems/prob019/
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:12]:
```suggestion
 * Quote taken from the [CSPLib](https://www.csplib.org/Problems/prob019/) website
```

File: backtracking/magic_sequence.cpp [hunk 10]
```diff
@@ -0,0 +1,146 @@
+/*
+ * @brief [Magic sequence](https://www.csplib.org/Problems/prob019/)
+ * implementation
+ *
+ * @details Solve the magic sequence problem with a backtraking
+ *
+ * "A magic sequence of length $n$ is a sequence of integers $x_0
+ *  \ldots x_{n-1}$ between $0$ and $n-1$, such that for all $i$
+ *  in $0$ to $n-1$, the number $i$ occurs exactly $x_i$ times in
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:9]:
```suggestion
 * "A magic sequence of length $n$ is a sequence of integers $x_0
 * \ldots x_{n-1}$ between $0$ and $n-1$, such that for all $i$
 * in $0$ to $n-1$, the number $i$ occurs exactly $x_i$ times in
```

File: backtracking/magic_sequence.cpp [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+/**
+ * @namespace magic_sequence
+ * @brief Functions for the [Magic sequence](https://www.csplib.org/Problems/prob019/) implementation
+ */
+namespace magic_sequence {
+using sequence_t = std::vector<unsigned int>;
+
+/**
+ * @brief print a magic sequence
+ *
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:39]:
```suggestion
```

File: backtracking/magic_sequence.cpp [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+ * @param s a magic sequence
+ */
+void print(const sequence_t& s) {
+    for (const auto& item : s) std::cout << item << " ";
+    std::cout << std::endl;
+}
+
+/**
+ * @brief Check if the sequence is magic
+ *
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:49]:
```suggestion
```

File: backtracking/magic_sequence.cpp [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        if (std::count(s.cbegin(), s.cend(), i) != s[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * @brief Sub-solutions filtering
+ *
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:66]:
```suggestion
```

File: backtracking/magic_sequence.cpp [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+ * @param s a magic sequence
+ * @param depth
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:68]: Missing documentation for this parameter.

File: backtracking/magic_sequence.cpp [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+ * @return false otherwise
+ */
+bool filtering(const sequence_t& s, unsigned int depth) {
+    return std::accumulate(s.cbegin(), s.cbegin() + depth,
+                           static_cast<unsigned int>(0)) <= s.size();
+}
+
+/**
+ * @brief Solve the Magic Sequence problem
+ *
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:79]:
```suggestion
```

File: backtracking/magic_sequence.cpp [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+    return true;
+}
+
+/**
+ * @brief Sub-solutions filtering
+ *
+ * @param s a magic sequence
+ * @param depth
+ * @return true if the sub-solution is valid
+ * @return false otherwise
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:70]:
```suggestion
 * @returns true if the sub-solution is valid
 * @returns false if the sub-solution is NOT valid
```

File: backtracking/magic_sequence.cpp [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+        }
+    }
+}
+
+}  // namespace magic_sequence
+}  // namespace backtracking
+
+/**
+ * @brief Self-test implementations
+ *
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:104]:
```suggestion
```

File: backtracking/magic_sequence.cpp [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+ * @returns void
+ */
+static void test() {
+    // test a valid magic sequence
+    backtracking::magic_sequence::sequence_t s_magic = {6, 2, 1, 0, 0,
+                                                        0, 1, 0, 0, 0};
+    assert(backtracking::magic_sequence::is_magic(s_magic));
+
+    // test a not valid magic sequence
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:113]:
```suggestion
    // test a non-valid magic sequence
```

File: backtracking/magic_sequence.cpp [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+ * @returns 0 on exit
+ */
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:134]:
```suggestion
```

File: backtracking/magic_sequence.cpp [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+int main() {
+    test();  // run self-test implementations
+
+    // solve magic sequences of size 2 to 11 and print the solutions
+    for (unsigned int i = 2; i < 12; i++) {
+        std::cout << "Solution for n = " << i << std::endl;
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:140]: You could add one-line documentation for each of these lines to be more understandable. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: backtracking/magic_sequence.cpp (hunk 20 above)
Comment (Jxtopher) [reply backtracking/magic_sequence.cpp:125]: Thank you for this improvement proposal.

File: backtracking/magic_sequence.cpp [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+    backtracking::magic_sequence::sequence_t s_magic = {6, 2, 1, 0, 0,
+                                                        0, 1, 0, 0, 0};
+    assert(backtracking::magic_sequence::is_magic(s_magic));
+
+    backtracking::magic_sequence::sequence_t s_not_magic = {5, 2, 1, 0, 0,
+                                                            0, 1, 0, 0, 0};
+    assert(!backtracking::magic_sequence::is_magic(s_not_magic));
+}
+
+int main() {
```
[the remark this answers is not in the document]
Comment (Jxtopher) [reply backtracking/magic_sequence.cpp:117]: Check

File: backtracking/magic_sequence.cpp [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+/**
+ * @namespace magic_sequence
+ * @brief Functions for the [Magic sequence](https://www.csplib.org/Problems/prob019/) implementation
+ */
+namespace magic_sequence {
+using sequence_t = std::vector<unsigned int>;
+
+/**
+ * @brief print a magic sequence
+ * @param s a magic sequence
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:38]:
```suggestion
 * @param s the magic sequence used to be printed
```

File: backtracking/magic_sequence.cpp [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+void print(const sequence_t& s) {
+    for (const auto& item : s) std::cout << item << " ";
+    std::cout << std::endl;
+}
+
+/**
+ * @brief Check if the sequence is magic
+ * @param s a magic sequence
+ * @return true if is a magic sequence
+ * @return false otherwise
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:49]:
```suggestion
 * @returns true if it's a magic sequence
 * @returns false if it's NOT a magic sequence
```

File: backtracking/magic_sequence.cpp [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+ * @param s a magic sequence
+ */
+void print(const sequence_t& s) {
+    for (const auto& item : s) std::cout << item << " ";
+    std::cout << std::endl;
+}
+
+/**
+ * @brief Check if the sequence is magic
+ * @param s a magic sequence
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:47]: Could you provide a better description of this parameter?

File: backtracking/magic_sequence.cpp [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * @brief Sub-solutions filtering
+ * @param s a magic sequence
+ * @param depth tree depth
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:64]: Could you provide a better description of these parameters?

File: backtracking/magic_sequence.cpp [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+ * @returns false if the sub-solution is NOT valid
+ */
+bool filtering(const sequence_t& s, unsigned int depth) {
+    return std::accumulate(s.cbegin(), s.cbegin() + depth,
+                           static_cast<unsigned int>(0)) <= s.size();
+}
+
+/**
+ * @brief Solve the Magic Sequence problem
+ * @param s a magic sequence
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:75]: Could you provide a better description of this parameter?

File: backtracking/magic_sequence.cpp [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+ * @param ret list of valid magic sequences
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:76]:
```suggestion
 * @param ret list of the valid magic sequences
```

File: backtracking/magic_sequence.cpp [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+ * @param depth depth in the tree
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:77]: Could you provide a better description of this parameter?

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: backtracking/magic_sequence.cpp [hunk 29]
```diff
@@ -0,0 +1,133 @@
+/*
+ * @brief [Magic sequence](https://www.csplib.org/Problems/prob019/) implementation
+ *
+ * @details Solve the magic sequence problem with a backtraking
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:4]:
```suggestion
 * @details Solve the magic sequence problem with backtracking
```

File: backtracking/magic_sequence.cpp [hunk 30]
```diff
@@ -0,0 +1,133 @@
...
+ *
+ * "A magic sequence of length $n$ is a sequence of integers $x_0
+ * \ldots x_{n-1}$ between $0$ and $n-1$, such that for all $i$
+ * in $0$ to $n-1$, the number $i$ occurs exactly $x_i$ times in
+ * the sequence. For instance, $6,2,1,0,0,0,1,0,0,0$ is a magic
+ * sequence since $0$ occurs $6$ times in it, $1$ occurs twice, etc."
+ * Quote taken from the [CSPLib](https://www.csplib.org/Problems/prob019/) website
+ *
+ * @author [Jxtopher](https://github.com/jxtopher)
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:13]:
```suggestion
 * @author [Jxtopher](https://github.com/Jxtopher)
```

File: backtracking/magic_sequence.cpp [hunk 31]
```diff
@@ -0,0 +1,133 @@
...
+#include <list>       /// for std::list
+#include <numeric>    /// for std::accumulate
+#include <vector>     /// for std::vector
+
+/**
+ * @namespace backtracking
+ * @brief Backtracking algorithms
+ */
+namespace backtracking {
+
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:28]:
```suggestion
```

File: backtracking/magic_sequence.cpp [hunk 32]
```diff
@@ -0,0 +1,133 @@
...
+/**
+ * @namespace magic_sequence
+ * @brief Functions for the [Magic sequence](https://www.csplib.org/Problems/prob019/) implementation
+ */
+namespace magic_sequence {
+using sequence_t = std::vector<unsigned int>;  ///< Definition of the sequence type
+
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:35]:
```suggestion
```

File: backtracking/magic_sequence.cpp [hunk 33]
```diff
@@ -0,0 +1,133 @@
...
+    for (const auto& item : s) std::cout << item << " ";
+    std::cout << std::endl;
+}
+
+/**
+ * @brief Check if the sequence is magic
+ * @param s working memory for the sequence
+ * @returns true if it's a magic sequence
+ * @returns false if it's NOT a magic sequence
+ *
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:50]:
```suggestion
```

File: backtracking/magic_sequence.cpp [hunk 34]
```diff
@@ -0,0 +1,133 @@
...
+bool filtering(const sequence_t& s, unsigned int depth) {
+    return std::accumulate(s.cbegin(), s.cbegin() + depth,
+                           static_cast<unsigned int>(0)) <= s.size();
+}
+
+/**
+ * @brief Solve the Magic Sequence problem
+ * @param s working memory for the sequence
+ * @param ret list of the valid magic sequences
+ * @param depth current depth in tree
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:77]:
```suggestion
 * @param depth current depth in the tree
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there; good work! 😄

File: backtracking/magic_sequence.cpp [hunk 35]
```diff
@@ -0,0 +1,135 @@
...
```
Comment (Panquesito7) [inline backtracking/magic_sequence.cpp:78]:
```suggestion
 * @param depth current depth in the tree
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — APPROVED** [review]
LGTM. Thank you for your patience, dedication, and contributions to our community! 😄 👍 🎉

Join our [Discord server](https://discord.gg/c7MnfGFGa6) to stay tuned about new announcements, engage with our community, help others with their problems, and much more.
You can also ask for the `Contributor` in the `#privs-role-requests` channel role once this PR is merged. 🙂

**Comment by Jxtopher** [comment]
Thank you for all the improvements proposed and this dynamic in the proposals.

**Review by ayaankhan98 — APPROVED** [review]

**Pull request merged**